Outlook activity datastream for o365 package

ritalwar opened the pull request:
- Enhancement


## Proposed commit message

Initial draft of the o365_metrics package with the `outlook_activity` data stream.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).
- [ ] I have verified that any added dashboard complies with Kibana's [Dashboard good practices](https://docs.elastic.dev/ux-guidelines/data-viz/dashboard-good-practices) 



## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues





## Screenshots

<!-- Optional
Add here screenshots presenting:
- Kibana UI forms presenting configuration options exposed by the integration
- dashboards with collected metrics or logs
-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/18948)
* Commit: 88d3147dc118186294e707f3cf4562d018bf7d8e

cc @ritalwar

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/18948","number":18948,"commit":"88d3147dc118186294e707f3cf4562d018bf7d8e"}],"number":18948}
buildkite-pr-comment-->

lucian-ioan commented on packages/o365_metrics/_dev/build/docs/README.md:
```diff
@@ -0,0 +1,44 @@
+# Microsoft Office 365 Metrics Integration
+
+This integration is to collect metrics for [Microsoft Office 365](https://learn.microsoft.com/en-us/graph/overview).
```
Maybe use a more descriptive introduction here.

```suggestion
This integration collects key metrics from [Microsoft Office 365](https://learn.microsoft.com/en-us/graph/overview) using the Graph API, providing insights into usage patterns and performance.
```

lucian-ioan commented on packages/o365_metrics/data_stream/outlook_activity/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,115 @@
+    azure.tenant_id: {{azure_tenant_id}}
+{{/if}}
+
+resource.url: https://graph.microsoft.com/v1.0/reports/getEmailActivityCounts(period='D7')
```
Consider not hardcoding and making this user configurable as certain parts of this endpoint (eg. v1.0) might change over time.

lucian-ioan commented on packages/o365_metrics/changelog.yml:
```diff
@@ -0,0 +1,6 @@
+# newer versions go on top
+- version: "0.0.1"
```
As discussed, keep this as `0.0.1-next` for now.

ritalwar replied:
I was checking the [documentation](https://github.com/elastic/integrations/blob/927d060c0d1d394c1c05f84520891eee4d780d6c/docs/developer_workflow_design_build_test_integration.md#remember-to-bump-up-the-version) and it seems we can use 0.1.0 based on what it [says](https://github.com/elastic/elastic-package/blob/main/docs/howto/use_package_storage_v2.md#prerelease-and-stable-version).

Also, using 0.0.1-next results in an error: `versions below 1.0.0 are considered technical previews. Please remove the prerelease tag (version: 0.0.1-next)`

lucian-ioan commented on packages/o365_metrics/manifest.yml:
```diff
@@ -0,0 +1,96 @@
+    title: Microsoft Office 365 Metrics
+    size: 216x216
+    type: image/svg+xml
+# screenshots:
+#   - src: /img/filebeat-o365-audit.png
+#     title: Office 365 Dashboard
+#     size: 1924x1409
+#     type: image/png
+#   - src: /img/filebeat-o365-azure-permissions.png
+#     title: Azure Permissions
+#     size: 2660x1030
+#     type: image/png
```
I'd remove this and replace later if dashboards are added.

lucian-ioan commented on packages/o365_metrics/data_stream/outlook_activity/fields/fields.yml:
```diff
@@ -0,0 +1,35 @@
+    - name: report_date
+      type: date
+      description: |
+        The specific date for which the report data applies..
```
The specific date for which the report data applies..
```suggestion
        The specific date for which the report data applies.
```

lucian-ioan commented on packages/o365_metrics/changelog.yml:
```diff
@@ -0,0 +1,6 @@
+  changes:
+    - description: Initial draft of the O365 metrics integration package.
+      type: enhancement
+      link: tbd
```
```suggestion
      link: https://github.com/elastic/integrations/pull/11937
```

lucian-ioan commented on packages/o365_metrics/changelog.yml:
```diff
@@ -0,0 +1,6 @@
+# newer versions go on top
+- version: "0.0.1"
+  changes:
+    - description: Initial draft of the O365 metrics integration package.
```
```suggestion
    - description: Add new data stream 'outlook_activity' for O365 metrics.
```

lucian-ioan reviewed (changes requested):
Some minor nits and suggestions, otherwise looks good.

ritalwar commented on packages/o365_metrics/data_stream/outlook_activity/_dev/test/pipeline/test-metrics.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "events": [
+      {
+          "outlookactivity": "{\"Meeting Created\":\"0\",\"Meeting Interacted\":\"\",\"Read\":\"\",\"Receive\":\"1\",\"Report Date\":\"2024-11-25\",\"Report Period\":\"7\",\"Send\":\"\",\"Report Refresh Date\":\"2024-11-30\"}"
+      }
+  ]
+
+}
```
Testing the pipeline with the input event should work fine

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/19144)
* Commit: 52e729374c1c1e5fc363c6a0e5290cfbe2b91272

### History
* :green_heart: [Build #18948](https://buildkite.com/elastic/integrations/builds/18948) succeeded 88d3147dc118186294e707f3cf4562d018bf7d8e

cc @ritalwar

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/19144","number":19144,"commit":"52e729374c1c1e5fc363c6a0e5290cfbe2b91272"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/18948","number":18948,"commit":"88d3147dc118186294e707f3cf4562d018bf7d8e"}],"number":19144}
buildkite-pr-comment-->

lucian-ioan reviewed (approved):
LGTM. I think we can ship this as the first one.

tommyers-elastic commented on packages/o365_metrics/data_stream/outlook_activity/fields/base-fields.yml:
```diff
@@ -0,0 +1,15 @@
+- name: data_stream.type
```
data_stream.* and @timestamp are in ECS, so aren't required here. where is `input.type` coming from?

muthu-mps replied:
@tommyers-elastic - The `data_stream.* `and `@timestamp` are mandatory fields that every package needs to have. This may be for the hardcoded values for constant_keywords like datastream.dataset. We cannot remove these fields as the package-spec adds mandatory field validation [here](https://github.com/elastic/package-spec/blob/5325b7e43d4b49cd6fdb90aa24705095911211e6/code/go/internal/validator/semantic/validate_required_fields.go#L16).

@ritalwar - Please verify the `input.type` field.

ritalwar replied:
It seems the `input.type` field is being sourced from the manifest file, as I removed it from this file, yet the field still appears.

ritalwar replied:
Yes, but for other fields, when I attempted to remove them, I encountered errors such as "expected field 'data_stream.dataset' with type 'constant_keyword'," among others. This indicates that these fields are mandatory.

tommyers-elastic replied:
https://github.com/elastic/package-spec/issues/843

muthu-mps replied:
@tommyers-elastic - Once the `package-spec` change is done to remove the restriction, Are we good to remove the base-fields from all the integration data streams?

tommyers-elastic replied:
maybe, we can address it when the change is made. i don't really see the need to remove all the existing mappings. it just means we don't need them in the future.

tommyers-elastic commented on packages/o365_metrics/data_stream/outlook_activity/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,38 @@
+      target_field: o365metrics.outlookactivity   
+  - script:
+      lang: painless
+      description: Replace spaces and dashes in field names under o365metrics.outlookactivity.
```
why do we need to do this? there aren't any dynamic fields are there?

ritalwar replied:
Yes, the fields are dynamic as we are fetching the complete API response, which for this API looks like:

```
{
  "Meeting Created": "0",
  "Meeting Interacted": "",
  "Read": "",
  "Receive": "9",
  "Report Date": "2024-11-24",
  "Report Period": "7",
  "Send": "",
  "Report Refresh Date": "2024-11-30"
}
```
Most fields are necessary, but since field names contain spaces, we need to remove them and convert them to lowercase to align with naming conventions.

tommyers-elastic replied:
oh i see, does the CEL input not provide a way to rename fields there?

ritalwar replied:
We could map field names in CEL and assign their values, similar to accessing map values, but that would involve hardcoding every field. Instead, I prefer fetching all fields dynamically, removing unnecessary ones using processors in the pipeline, and applying a single generic script for naming.

lalit-satapathy replied:
Can this thread be closed?

ritalwar commented on packages/o365_metrics/data_stream/outlook_activity/fields/fields.yml:
```diff
@@ -0,0 +1,35 @@
+- name: o365metrics.outlookactivity
+  type: group
+  fields:
+    - name: meeting_created
```
I tried to follow the conventions by suffixing `.count`, but the events fail to index, dropping with errors like: ```*"object mapping for [o365metrics.outlookactivity.meeting_interacted] tried to parse field [meeting_interacted] as object, but found a concrete value."* ```
This could be a data type mapping issue, but even after trying different data type mappings, the issue persists. 
I think it might be best to create a separate issue to address this as a refinement and proceed with merging this initial PR. What do you suggest?

ritalwar commented on packages/o365_metrics/data_stream/outlook_activity/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,115 @@
+    state.with(
+      request(
+        "GET",
+        state.url + "/reports/getEmailActivityCounts(period='D7')"
```
Good point! I've added the period configuration in the manifest, made it configurable, and updated the code accordingly.

ritalwar commented on packages/o365_metrics/data_stream/outlook_activity/fields/fields.yml:
```diff
@@ -0,0 +1,35 @@
+- name: o365metrics.outlookactivity
```
We plan to address TSDB-related changes later, once a few datastreams are merged. We've also asked for help to perform some service-related activities to generate richer API responses, which will help us implement these changes effectively.

lalit-satapathy replied:
Is units done and only dimension and metric types pending?

ritalwar replied:
I have suffixed the names with `.count` where applicable. However, the unit value `count` is not allowed due to validation constraints, which specify that the unit must be one of the following: `"byte"`, `"percent"`, `"d"`, `"h"`, `"m"`, `"s"`, `"ms"`, `"micros"`, or `"nanos"`.

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/19410)
* Commit: b985b28834a9bbf71b12a509115d0e8a375d17b9

### History
* :green_heart: [Build #19354](https://buildkite.com/elastic/integrations/builds/19354) succeeded 0d16f81adae6c0bf160d0d40693a36971a4ed7a5
* :green_heart: [Build #19144](https://buildkite.com/elastic/integrations/builds/19144) succeeded 52e729374c1c1e5fc363c6a0e5290cfbe2b91272
* :green_heart: [Build #18948](https://buildkite.com/elastic/integrations/builds/18948) succeeded 88d3147dc118186294e707f3cf4562d018bf7d8e

cc @ritalwar

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/19410","number":19410,"commit":"b985b28834a9bbf71b12a509115d0e8a375d17b9"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/19354","number":19354,"commit":"0d16f81adae6c0bf160d0d40693a36971a4ed7a5"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/19144","number":19144,"commit":"52e729374c1c1e5fc363c6a0e5290cfbe2b91272"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/18948","number":18948,"commit":"88d3147dc118186294e707f3cf4562d018bf7d8e"}],"number":19410}
buildkite-pr-comment-->

lalit-satapathy commented on packages/o365_metrics/manifest.yml:
```diff
@@ -0,0 +1,87 @@
+name: o365_metrics
+title: Microsoft Office 365 Metrics
+version: "0.1.0"
+description: Collect metrics from Microsoft Office 365 with Elastic Agent.
```
should we add _(under-development)_ at the end?

ritalwar replied:
Updated, and it looks like:
<img width="392" alt="Screenshot 2024-12-16 at 10 56 43 AM" src="https://github.com/user-attachments/assets/2e5b1454-b143-4b8c-a79f-b95e2edd8ff6" />

muthu-mps commented on packages/o365_metrics/data_stream/outlook_activity/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,59 @@
+processors:
+  - set:
+      field: ecs.version
+      value: "8.11.0"
```
The `build.yml` sets `8.16` as reference ECS version. Do you think we need to update the ECS version to the latest one here?

ritalwar replied:
What’s the concern? We now have support for ECS 8.16, and many are also already migrating. It was also discussed on Slack to use this version for new package development.

muthu-mps replied:
No issues in using 8.16 ECS version, Why are we setting ecs.version to 8.11 here? Is there any specific reason to set the processor version to 8.11?

ritalwar replied:
Oh, I misread earlier. I thought I had updated it to 8.16 in both places. I've corrected it now. Thanks!

lalit-satapathy commented:
- For the field naming we have now as _o365metrics.*_. One alternative is _o365.metrics_. Any suggestion, which is preferable? Also the O365 audit package has fields named as: _o365.audit.*_, if we want to be consistent to that.
- For data stream naming, also lets define a consistent scheme between these choices and something which works for all future data_stream names:
  - outlookactivity vs.
  - outlook.activity vs.
  - outlook_activity
(Also please note it may be confusing to create a future data_stream name for _m365.active_users_ since, o365 is already being used in the beginning).

CC: @lucian-ioan

ritalwar commented:
> * For the field naming we have now as _o365metrics.*_. One alternative is _o365.metrics_. Any suggestion, which is preferable? Also the O365 audit package has fields named as: _o365.audit.*_, if we want to be consistent to that.

Yes, I think it's better to maintain consistency, so I've updated the naming to `o365.metrics`.

> * For data stream naming, also lets define a consistent scheme between these choices and something which works for all future data_stream names:
>   
>   * outlookactivity vs.
>   * outlook.activity vs.
>   * outlook_activity

For field naming, I have updated it to `outlook.activity`, following the structure `o365.metrics.outlook.activity.*`. However, for the datastream name, the naming convention validation does not allow the use of `outlook.activity`. To maintain clarity, I suggest sticking with `outlook_activity` for the datastream name.

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/19530)
* Commit: 6028d05d3a1734330415e69512748c726dd4bb3d

### History
* :green_heart: [Build #19410](https://buildkite.com/elastic/integrations/builds/19410) succeeded b985b28834a9bbf71b12a509115d0e8a375d17b9
* :green_heart: [Build #19354](https://buildkite.com/elastic/integrations/builds/19354) succeeded 0d16f81adae6c0bf160d0d40693a36971a4ed7a5
* :green_heart: [Build #19144](https://buildkite.com/elastic/integrations/builds/19144) succeeded 52e729374c1c1e5fc363c6a0e5290cfbe2b91272
* :green_heart: [Build #18948](https://buildkite.com/elastic/integrations/builds/18948) succeeded 88d3147dc118186294e707f3cf4562d018bf7d8e

cc @ritalwar

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/19530","number":19530,"commit":"6028d05d3a1734330415e69512748c726dd4bb3d"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/19410","number":19410,"commit":"b985b28834a9bbf71b12a509115d0e8a375d17b9"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/19354","number":19354,"commit":"0d16f81adae6c0bf160d0d40693a36971a4ed7a5"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/19144","number":19144,"commit":"52e729374c1c1e5fc363c6a0e5290cfbe2b91272"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/18948","number":18948,"commit":"88d3147dc118186294e707f3cf4562d018bf7d8e"}],"number":19530}
buildkite-pr-comment-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/19531)
* Commit: 70500e2d9d698c4103369918a9218fb43d35f769

### History
* :green_heart: [Build #19530](https://buildkite.com/elastic/integrations/builds/19530) succeeded 6028d05d3a1734330415e69512748c726dd4bb3d
* :green_heart: [Build #19410](https://buildkite.com/elastic/integrations/builds/19410) succeeded b985b28834a9bbf71b12a509115d0e8a375d17b9
* :green_heart: [Build #19354](https://buildkite.com/elastic/integrations/builds/19354) succeeded 0d16f81adae6c0bf160d0d40693a36971a4ed7a5
* :green_heart: [Build #19144](https://buildkite.com/elastic/integrations/builds/19144) succeeded 52e729374c1c1e5fc363c6a0e5290cfbe2b91272
* :green_heart: [Build #18948](https://buildkite.com/elastic/integrations/builds/18948) succeeded 88d3147dc118186294e707f3cf4562d018bf7d8e

cc @ritalwar

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/19531","number":19531,"commit":"70500e2d9d698c4103369918a9218fb43d35f769"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/19530","number":19530,"commit":"6028d05d3a1734330415e69512748c726dd4bb3d"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/19410","number":19410,"commit":"b985b28834a9bbf71b12a509115d0e8a375d17b9"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/19354","number":19354,"commit":"0d16f81adae6c0bf160d0d40693a36971a4ed7a5"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/19144","number":19144,"commit":"52e729374c1c1e5fc363c6a0e5290cfbe2b91272"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/18948","number":18948,"commit":"88d3147dc118186294e707f3cf4562d018bf7d8e"}],"number":19531}
buildkite-pr-comment-->

lalit-satapathy reviewed (approved):
Early development version. Approving to enable successive PR merge.
